src/dune_engine/dep.ml

```diff
@@ -315,44 +314,36 @@ module Set = struct
     file_selector
       (File_selector.create ~dir File_selector.Predicate_with_id.false_)
 
-  module Source_tree_map_reduce =
-    Source_tree.Dir.Make_map_reduce (Memo) (Monoid.Union (M))
-
-  let source_tree dir =
-    let prefix_with, dir = Path.extract_build_context_dir_exn dir in
-    Source_tree.find_dir dir >>= function
-    | None -> Memo.return empty
-    | Some dir ->
-      Source_tree_map_reduce.map_reduce dir ~traverse:Sub_dirs.Status.Set.all
-        ~f:(fun dir ->
-          let files = Source_tree.Dir.files dir in
-          let path =
-            Path.append_source prefix_with (Source_tree.Dir.path dir)
-          in
+  module Source_tree (Union : sig
+    type 'a result
+
+    val union_all :
+      Path.t -> f:(path:Path.t -> files:Filename.Set.t -> t) -> t result
+  end) =
+  struct
+    let files dir =
+      Union.union_all dir ~f:(fun ~path ~files ->
           match String.Set.is_empty files with
-          | true -> Memo.return (singleton (dir_without_files_dep path))
+          | true -> singleton (dir_without_files_dep path)
           | false ->
             let paths =
               String.Set.fold files ~init:Path.Set.empty ~f:(fun fn acc ->
                   Path.Set.add acc (Path.relative path fn))
             in
-            Memo.return (add_paths empty paths))
+            add_paths empty paths)
```

Comment:
The above `fold` followed by another `fold` in `add_paths` look a little silly. Can we fuse them and get rid of an intermediate `Path.Set.t`?

(Sorry, I keep complaining about the code that existed before this PR but I feel like it's a good opportunity to clean it up a little.)